File: SpaceSim/Simulation/Market.cs
using System;
using System.Collections.Generic;
namespace Toy.Simulation
{
    public class Market
    {
        double elasticity;
        double maxPrice;
        int supply;

        public Market(Good good, double maxPrice, double elasticity, int startingSupply) {
            this.maxPrice = maxPrice;
            this.elasticity = elasticity;
            this.supply = startingSupply;
        }

        public double CurrentPrice() {
            //P = a − bQ, a - the price at which none of the good is desired, b the elasticity, Q the quantity available
            return Math.Max(maxPrice - (elasticity * supply), 0);
        }

        public double Buy() {
            if (supply == 0) {
                throw new Exception("No available sellers");
            }
            double price = CurrentPrice();
            supply--;
            return price;
        }

        public double Sell() {
            double price = CurrentPrice();
            supply++;
            return price;
        }

        public double Buy(int quantity) {
            if (supply >= quantity)
            {
                double startingPrice = CurrentPrice();
                supply -= quantity;
                double endingPrice = CurrentPrice();

                return (startingPrice + (endingPrice + 1)) / 2.0;
            }
            else
            {
                throw new Exception("Insufficient goods");
            }
        }

        public double Sell(int quantity)
        {
            double revenue = 0.0;
            for (int i = 0; i < quantity; i++)
            {
                revenue += Sell();
            }
            return revenue;
        }
    }
}

File: simulation/market.py
﻿import math
from good import Good

class Market:
    def __init__(self, good: Good, max_price, elasticity, starting_supply):
        self.max_price = max_price
        self.elasticity = elasticity
        self.supply = starting_supply

    def current_price(self):
        # P = a − bQ, a - the price at which none of the good is desired, b the elasticity, Q the quantity available
        return max(self.max_price - (self.elasticity * self.supply), 0)

    def buy_one(self):
        if self.supply is 0:
            raise Exception("No available sellers")
        price = self.current_price()
        self.supply = self.supply - 1
        return price

    def sell_one(self):
        price = self.current_price()
        self.supply = self.supply + 1
        return price

    def buy(self, quantity):
        if self.supply < quantity:
            raise Exception("Insufficient goods")

        startingPrice = self.current_price()
        self.supply = self.supply - quantity
        endingPrice = self.current_price()
        return (startingPrice + (endingPrice + 1)) / 2.0

    def sell(self, quantity):
        revenue = 0
        for i in range(quantity):
            revenue = revenue + self.sell_one()
        return revenue
